datamonster_api/lib/datamonster.py

```diff
@@ -199,61 +237,50 @@ def get_data(
+        resp = self.client.get(url, headers, stream=True)
+        split_columns = datasource.get_details()["splitColumns"]
+        return self._avro_to_df(resp.content, split_columns)
 
-        return self._avro_to_df(resp)
-
-    def get_datasource_details(self, datasource_id):
-        """Get details (metadata) for the given datasource
+    def _avro_to_df(self, avro_buffer, split_columns):
+        """Read an avro structure into a dataframe
 
-        :param datasource_id: The ID of the datasource for which we get the details
-        :return: dictionary object with the datasource details
+        Transforms dates and columns to a stanard and agreed upon format
         """
-        path = self._get_datasource_path(datasource_id)
-        return self.client.get(path)
-
-    def _get_datasource_path(self, datasource_id):
-        return "{}/{}".format(self.datasource_path, datasource_id)
-
-    def _datasource_result_to_object(self, datasource, has_details=False):
-        ds_inst = Datasource(
-            datasource["id"],
-            datasource["name"],
-            datasource["category"],
-            datasource["uri"],
-            self,
-        )
-
-        if has_details:
-            ds_inst.set_details(datasource)
 
-        return ds_inst
+        def parse_row(row, data_column):
+            return {
+                "value": row[data_column],
+                "start_date": row["period_start"],
+                "end_date": row["period_end"],
+                "dimensions": {
+                    split_key: row[split_key] for split_key in split_columns
+                },
+            }
+
+        reader = fastavro.reader(six.BytesIO(avro_buffer))
+        metadata = reader.schema["structure"]
+        if not metadata:
+            raise DataMonsterError(
+                "DataMonster does not currently support this request"
+            )
 
-    def _avro_to_df(self, avro_buffer):
-        """Read an avro structure into a dataframe
-        """
-        fp = six.BytesIO(avro_buffer)
-        reader = fastavro.reader(fp)
         records = [r for r in reader]
-        df = pandas.DataFrame.from_records(records)
 
-        if len(df) == 0:
-            return df
+        if not records:
+            return pandas.DataFrame.from_records(records)
 
-        # Convert date columns to datetime64 columns
-        df["upperDate"] = df["upperDate"].astype("datetime64[ns]")
-        df["lowerDate"] = df["lowerDate"].astype("datetime64[ns]")
+        records = [parse_row(row, metadata["value"][0]) for row in records]
 
+        df = pandas.DataFrame.from_records(records)
+        df["start_date"] = df["start_date"].astype("datetime64[ns]")
+        df["end_date"] = df["end_date"].astype("datetime64[ns]")
         # Create the timespan. Note we add 1 day because both dates are inclusive
-        df["time_span"] = df["upperDate"] - df["lowerDate"] + timedelta64(1, "D")
+        df["end_date"] -= datetime.timedelta(days=1)
```

Comment:
If you reverse these 2 lines you don't have to add back in the day for the time_span math. Comment could then be "Change end_date to inclusive" or something like that.